### PR TITLE
input_common: joycon: Remove Magic numbers from common protocol

### PR DESCRIPTION
**Pull request by german77**
Same deal as previous PR. Removes the usage of vectors and return proper structs. Eliminating all magic numbers in the process.

### CODE REVIEW

File: src/input_common/helpers/joycon_protocol/joycon_types.h [hunk 1]
```diff
@@ -623,6 +636,53 @@ struct RingStatus {
...
+
+#pragma pack(push, 1)
+struct ReadSpiPacket {
+    SpiAddress spi_address;
+    INSERT_PADDING_BYTES(0x2);
+    u8 size;
+};
+static_assert(sizeof(ReadSpiPacket) == 0x5, "ReadSpiPacket is an invalid size");
+
+struct SubCommandResponce {
```
Comment (liamwhite) [inline src/input_common/helpers/joycon_protocol/joycon_types.h:663]:
```suggestion
struct SubCommandResponse {
```

File: src/input_common/helpers/joycon_protocol/common_protocol.cpp [hunk 2]
```diff
@@ -120,61 +119,72 @@ DriverResult JoyconCommonProtocol::SendSubCommand(SubCommand sc, std::span<const
...
+    constexpr std::size_t HeaderSize = 5;
     constexpr std::size_t MaxTries = 10;
-    const auto size = output.size();
     std::size_t tries = 0;
-    std::array<u8, 5> buffer = {0x00, 0x00, 0x00, 0x00, static_cast<u8>(size)};
-    std::vector<u8> local_buffer{};
-
-    buffer[0] = static_cast<u8>(static_cast<u16>(addr) & 0x00FF);
-    buffer[1] = static_cast<u8>((static_cast<u16>(addr) & 0xFF00) >> 8);
+    SubCommandResponce responce{};
```
Comment (liamwhite) [inline src/input_common/helpers/joycon_protocol/common_protocol.cpp:163]:
```suggestion
    SubCommandResponce response{};
```

File: src/input_common/helpers/joycon_protocol/joycon_types.h [hunk 3]
```diff
@@ -623,6 +636,53 @@ struct RingStatus {
...
+        SpiAddress spi_address;              // Reply from SPI_FLASH_READ subcommand
+        ExternalDeviceId external_device_id; // Reply from GET_EXTERNAL_DEVICE_INFO subcommand
+        DeviceInfo device_info;              // Reply from REQ_DEV_INFO subcommand
+    };
+    u8 crc; // This is never used
+};
+static_assert(sizeof(SubCommandResponce) == 0x40, "SubCommandResponce is an invalid size");
+#pragma pack(pop)
+
+struct MCUCommandResponce {
```
Comment (liamwhite) [inline src/input_common/helpers/joycon_protocol/joycon_types.h:677]:
```suggestion
struct MCUCommandResponse {
```

File: src/input_common/helpers/joycon_protocol/joycon_types.h [hunk 4]
```diff
@@ -623,6 +636,53 @@ struct RingStatus {
     s16 min_value;
 };
 
+struct VibrationPacket {
+    OutputReport ouput_report;
```
Comment (behunin) [inline src/input_common/helpers/joycon_protocol/joycon_types.h:640]:
```suggestion
    OutputReport output_report;
```

File: src/input_common/helpers/joycon_protocol/joycon_types.h [hunk 5]
```diff
@@ -623,6 +636,53 @@ struct RingStatus {
...
+    u8 packet_counter;
+    std::array<u8, 0x8> vibration_data;
+};
+static_assert(sizeof(VibrationPacket) == 0xA, "VibrationPacket is an invalid size");
+
+struct SubCommandPacket {
+    OutputReport ouput_report;
```
Comment (behunin) [inline src/input_common/helpers/joycon_protocol/joycon_types.h:647]:
```suggestion
    OutputReport output_report;
```

**Review by behunin — COMMENTED** [review]
Spelling

File: src/input_common/helpers/joycon_protocol/common_protocol.h [hunk 6]
```diff
@@ -57,22 +57,30 @@ class JoyconCommonProtocol {
      * Sends data to the joycon device
      * @param buffer data to be send
      */
-    DriverResult SendData(std::span<const u8> buffer);
+    DriverResult SendRawData(std::span<const u8> buffer);
+
+    template <typename Output>
+    requires std::is_trivially_copyable_v<Output> DriverResult SendData(const Output& output) {
```
Comment (behunin) [inline src/input_common/helpers/joycon_protocol/common_protocol.h:63]:
```suggestion
    requires std::is_trivially_copyable_v<Output>
    DriverResult SendData(const Output& output) {
```

Comment (liamwhite) [reply src/input_common/helpers/joycon_protocol/common_protocol.h:63]: Not compatible with clang-format 11

Comment (german77) [reply src/input_common/helpers/joycon_protocol/common_protocol.h:63]: We need to update ci clang for this. It's annoying.

**Review by liamwhite — APPROVED** [review]

File: src/input_common/helpers/joycon_protocol/common_protocol.h (hunk 6 above)
Comment (behunin) [reply src/input_common/helpers/joycon_protocol/common_protocol.h:63]: The inspiration for #9705
No more annoyances with requires formatting :)

**Pull request merged**